package/MDAnalysis/analysis/leaflet.py

```diff
@@ -322,3 +322,255 @@ def optimize_cutoff(universe, select, dmin=10.0, dmax=20.0, step=0.5,
+    """
+
+    def __init__(self, universe, select_protein: str='protein', 
+                 select_residues: str='all', select_headgroup: str='name PO4',
+                 n_leaflets: int=2, count_by_attr:str ='resnames', 
+                 enrichment_cutoff: float=6, delta: float=0.5,
+                 compute_headgroup_only: bool=True, **kwargs):
+        super(LipidEnrichment, self).__init__(universe.universe.trajectory, 
+                                              **kwargs)
+        self.box = universe.dimensions
+        self.protein = universe.select_atoms(select_protein)
+        self.residues = universe.select_atoms(select_residues).residues
+        self.n_residues = len(self.residues)
+        self.headgroups = self.residues.atoms.select_atoms(select_headgroup)
+        if n_leaflets < 1:
+            raise ValueError('Must have at least one leaflet')
+        self.n_leaflets = n_leaflets
+        self.delta = delta
+        self.compute_headgroup_only = compute_headgroup_only
+        self.cutoff = enrichment_cutoff
+        self.leaflets = []
+        self.leaflets_summary = []
+
+        # process this a bit to remove errors like "alt_locs"
+        attrname = _TOPOLOGY_ATTRNAMES[count_by_attr.lower().replace('_', '')]
+        self.attrname = _TOPOLOGY_ATTRS[attrname].attrname
+
+    def _prepare(self):
+        try:
+            import sklearn.cluster as skc
+        except ImportError:
+            raise ImportError('scikit-learn is required to use this analysis '
+                              'but is not installed. Install it with `conda '
+                              'install scikit-learn` or `pip install '
+                              'scikit-learn`.') from None
+
+        if self.n_leaflets > 1:
+            # determine leaflets by clustering center-of-geometry
+            coms = self.headgroups.center_of_geometry(compound='residues')
+            pdist = distances.distance_array(coms, coms, box=self.box)
+            psim = np.exp(-pdist**2/(2*(self.delta**2)))
+            sc = skc.SpectralClustering(n_clusters=self.n_leaflets,
+                                        affinity='precomputed_nearest_neighbors')
+            clusters = sc.fit_predict(pdist)
+            leaflets = []
+            for i in range(self.n_leaflets):
+                leaflets.append([])
+            for res, i in zip(self.residues, clusters):
+                leaflets[i].append(res)
+            self.leaflet_residues = lres = [np.array(x).sum() 
+                                            for x in leaflets]
+            self.leaflet_residues.sort(key=lambda x: x.center_of_geometry()[-1],
+                                       reverse=True)
+            self.leaflet_headgroups = [(self.headgroups & r.atoms) 
+                                       for r in lres]
+        else:
+            self.leaflet_residues = [self.residues]
+            self.leaflet_headgroups = [self.headgroups]
```

Comment:
If you can make the API similar to the existing LeafletFinder then one could also just add a `method=...` kwarg to LeafletFinder and either use the classic one or the spectral_clustering.

Reply:
> In any case, why hide it? Also, then you can test it independently.

I guess it felt very short to be its own class, but I can move it into LeafletFinder. 

> What's the difference between the build-in spectral-clustering leafletfinder and the old one? What's the advantage?

LeafletFinder constructs a networkx graph from a matrix of contacts, with a default cutoff of `cutoff=15`. The groups it returns are every connected component in the graph, sorted by size. It means you should seriously consider your `cutoff` parameter in case you accidentally make little lipid islands, or connect all the lipids; this directly affects whether the `n` biggest groups are true representations of your leaflets.

The spectral clustering method just clusters a similarity matrix into `n` leaflets, where you choose `n`. There's a `delta` parameter that indirectly affects the clustering because it's used to transform the distance matrix into the similarity matrix, but I haven't found the need to change it for the simulations we've used it on. I like this method because I have to do less thinking and I'm guaranteed to get `n` leaflets. If, e.g., cholesterol is floating around in the middle of flip-flopping it's guaranteed to get assigned to one of the leaflets. 

> Is the algorithm your own or from the paper? Would be worthwhile making it citable.

The leaflet-finding is just spectral clustering; the DEI formula is from the paper. I am thinking of putting in a soft potential to weight the distances of the lipids from the protein, though, as I am concerned that the hard cutoff + skipping frames neglects the transient movements of low-population lipids in a way that has an unrepresentatively large effect on the final index.

Reply:
I should mention that I did try some different `delta` values under 1 and it didn't seem to have any effect.

Reply:
The similarity matrix clustering of distances is so general that it could just live in the distances module. It can be easily applied to contacts, salt bridges, h-bonds... Then you just call it here with the headgroup COM positions as input.

There is probably a way to speed it up with capped distances because your similarity matrix does not really care about values beyond a few `delta`. 

@ianmkenney and @VOD555 might be interested, too.

Reply:
If it works well, it would be beneficial to expose the spectral leaflet finder. At the very least, it would allow to be consistent in the definition of the leaflets between analyses.